thorlcr/master/thmastermain.cpp

```diff
@@ -441,9 +441,67 @@ bool checkClusterRelicateDAFS(IGroup &grp)
 
 static bool auditStartLogged = false;
 
+class CThorEndHandler : public CSimpleInterface, implements IThreaded
+{
+    CThreaded threaded;
+    std::atomic<unsigned> timeout{30000};
+    std::atomic<bool> started{false};
+    std::atomic<bool> stopped{false};
+    Semaphore sem;
+public:
+    CThorEndHandler() : threaded("CThorEndHandler")
+    {
+        threaded.init(this);
+    }
+    ~CThorEndHandler()
+    {
+        stop();
+        threaded.join(timeout);
+    }
+    void start()
+    {
+        if (!started)
+        {
+            started = true;
+            sem.signal();
+        }
+    }
+    void start(unsigned timeoutsecs)
+    {
+        if (!started)
+        {
+            timeout = timeoutsecs * 1000;
+            start();
+        }
+    }
+    void stop()
+    {
+        if (!stopped)
+        {
+            stopped = true;
+            sem.signal();
+        }
+    }
+    virtual void threadmain() override
+    {
+        // wait to be signalled to start timer
+        sem.wait();
+        if (stopped)
+            return;
+        if (!sem.wait(timeout))
+            raise(SIGKILL);
+    }
+};
+CThorEndHandler thorendhdlr;
+
 static bool firstCtrlC = true;
 bool ControlHandler(ahType type)
 {
+    // MCK - NOTE: this routine may make calls to non-async-signal safe functions
+    //             (such as malloc) that really should not be made if we are called
+    //             from a signal handler - start end handler timer to always end
+    thorendhdlr.start(120);
```

Comment:
Is the reason for pre-constructing and pre-starting the thread on line 495, to avoid creating a new thread here inside a signal handler?

I think so and makes sense, but the pre-existing code is also not safe (it calls malloc for example when calling mp cancel()). So all (or most) this code should move out of the signal handler I suspect and be called on a thread.

Reply:
Correct, we really should move the bulk of the signal handler code into a separate thread and the handler just signals that thread to go on.  But that is a separate and larger issue.  I added the thorendhdlr.start(120) here as it seemd good insurance for Thor to always end.